File: questions/378. Kth Smallest Element in a Sorted Matrix/k.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        queue = []
        m, n, res = len(matrix), len(matrix[0]), 0
        for i in range(min(m, k)):
            heapq.heappush(queue,(matrix[i][0], i, 0))
        while k > 0:
            num, row, col = heapq.heappop(queue)
            res = num
            if col + 1 < n:
                heapq.heappush(queue, (matrix[row][col + 1], row, col + 1))
            k -= 1
        return num
        

File: questions/378. Kth Smallest Element in a Sorted Matrix/k.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int m = matrix.length, n = matrix[0].length, res = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)->a[0]-b[0]);
        for(int i = 0; i < Math.min(k, m); i++){
            pq.add(new int[]{matrix[i][0], i, 0});
        }
        while(k > 0){
            int[] t = pq.poll();
            res = t[0];
            if(t[2] + 1 < n){
                pq.add(new int[]{matrix[t[1]][t[2] + 1], t[1], t[2] + 1});
            }
            k -= 1;
        }
        return res;
    }
}
